proposed/0000-Unified-structured-data.md

```diff
@@ -0,0 +1,128 @@
+
+##Expected outcome
+
+
+It is expected removing Transaction Managers from network will reduce complexity, code and increase security on the network, whilst allowing a greater degree of flexibility. These types now allow the users of this network to create their own data types and structures to be securely managed. This is hoped to allow many new types of application to exist.
+
+# Detailed design
+
+The design entails reducing all StructuredData types to a single type, therefore it should be able to be recognised by the network as StructuredData and all such sub-types handled exactly in the same manner. 
+
+##StructuredData
+
+```
+struct StructuredData {
+tag_type : TagType, // 8 Bytes ?
+identifier : NameType // 64Bytes (i.e. SHA512 Hash)
+data : mut Vec<u8>, // in many cases this is encrypted
+owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of owners)
+version : mut u64, // incrementing (deterministic) version number
+previous_owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of
+owners) only required when owners change 
+signature : mut Vec<Signature> // signs the `mut` fields above // 32 bytes (using e25519 sig)
+}
+```
+
+Fixed (immutable fields)
+- tag_type
+- identifier
+
+##Validation
+
+- To confirm name (storage location on network) we SHA512(tag_type + identifier). As these are much
+  smaller than the hash it prevents flooding of a network location.
+- To validate data we confirm signature using hash of (tag_type + version) as nonce. Initial `Put`
+  does not require this signature, but does require the owner contain a value.
+- If previous owners is set then the signature is confirmed using those keys (allows owenr
+  change/transfer etc.). In this case the previous owners is only required on first update and may be
+  remove in next version if the owners are not changed again.
+- To confirm sender of any `Put` (store or overwrite) then we check the signature of sender using same mechanism. For multiple senders we confirm at least 50% of owners have signed the request for `Put`
+
+When `Put` on the network this type is `StructuredData` with a subtype field. The network ignores this subtype
+except for collisions. No two data types with the same name and type can exist on the network.
+
+These types are stored in a disk based storage mechanism such as `Btree` at the NaeManagers (DataManagers) responsible for the area of the network at that `name`.
+
+These types will be limited to 100kB in size (as Immutable Chunks are also limited to 1Mb) which is for the time being a magic number.
+
+If a client requires these be larger than 100kB then the data component will contain a (optionally encrypted) datamap to be able to retrieve chunks of the network.
```

Comment:
This is just for seeking confirmation. Say the user wants versioning of DirectoryListing. Now with every update name of new datamaps (ImmutableData) are appended to the existing ones in the StructuredData. If this grows beyond 100KB, the last 64bytes Name will be the name of the new StructuredData (instead of name of datamap/ImmutableData) which will contain further versions (sort of chaining). There can be many ways to do it because `data` field completely belongs to client. So can any rule be put there or there needs to be a standardized process for this. Eg.,
One way is `data` = `serialised (Option<NameType of previous StructuredData>, Vec<NameTypes of DataMaps>, Option<NameType of Next StructuredData>)` .. So does client have flexibility to do all these tricks?